analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/CalculationsShouldNotOverflowBase.cs

```diff
@@ -18,12 +18,82 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
+using System.Numerics;
+using SonarAnalyzer.SymbolicExecution.Constraints;
+
 namespace SonarAnalyzer.SymbolicExecution.Roslyn.RuleChecks;
 
 public abstract class CalculationsShouldNotOverflowBase : SymbolicRuleCheck
 {
     protected const string DiagnosticId = "S3949";
-    protected const string MessageFormat = "This calculation is {0} to overflow the maximum value of '{1}'.";
+    protected const string MessageFormat = "This calculation is {0} to {1} value of '{2}'.";
     private const string MessageLikely = "likely";
     private const string MessageGuaranteed = "guaranteed";
+    private const string MessageUnderflow = "underflow the minimum";
+    private const string MessageOverflow = "overflow the maximum";
+
+    protected override ProgramState PostProcessSimple(SymbolicContext context)
+    {
+        if (context.Operation.Instance is var operation
+            && CanOverflow(context.State, operation)?.Constraint<NumberConstraint>() is { } number
+            && Min(operation.Type) is { } typeMin
+            && Max(operation.Type) is { } typeMax)
+        {
+            if (number.Max < typeMin)
+            {
+                ReportIssue(operation, MessageGuaranteed, MessageUnderflow, typeMax);
```

Comment:
All message variants needs to be asserted
```suggestion
                ReportIssue(operation, MessageGuaranteed, MessageUnderflow, typeMin);
```